tachyon/zk/plonk/halo2/argument.h

```diff
@@ -129,23 +129,40 @@ class Argument {
   // of instance evaluation-formed columns. (a.k.a. Batch IFFT)
   static std::vector<std::vector<Poly>> GenerateInstancePolys(
       ProverBase<PCS>* prover,
-      std::vector<std::vector<Evals>> instance_columns_vec) {
-    return base::Map(instance_columns_vec,
-                     [prover](const std::vector<Evals>& instance_columns) {
-                       return base::Map(
-                           instance_columns,
-                           [prover](const Evals& instance_column) {
-                             if constexpr (PCS::kQueryInstance) {
-                               prover->CommitAndWriteToProof(instance_column);
-                             } else {
-                               for (size_t i = 0; i < prover->pcs().N(); ++i) {
-                                 CHECK(prover->GetWriter()->WriteToTranscript(
-                                     *instance_column[i]));
-                               }
-                             }
-                             return prover->domain()->IFFT(instance_column);
-                           });
-                     });
+      const std::vector<std::vector<std::vector<F>>>& instance_3d_vec) {
+    CHECK_GT(instance_3d_vec.size(), size_t{0});
+    size_t num_circuit = instance_3d_vec.size();
+    size_t num_instance_columns = instance_3d_vec[0].size();
+    if constexpr (PCS::kSupportsBatchMode && PCS::kQueryInstance) {
+      size_t num_commitment = num_circuit * num_instance_columns;
+      prover->pcs().SetBatchMode(num_commitment);
+    }
+
+    std::vector<std::vector<Poly>> instance_polys_vec;
+    instance_polys_vec.reserve(num_circuit);
+    for (size_t i = 0; i < num_circuit; ++i) {
+      const std::vector<std::vector<F>>& instance_columns = instance_3d_vec[i];
+      std::vector<Poly> instance_polys;
+      instance_polys.reserve(num_instance_columns);
+      for (size_t j = 0; j < num_instance_columns; ++j) {
+        const std::vector<F>& instances = instance_columns[j];
+        if constexpr (PCS::kQueryInstance) {
```

Comment:
```suggestion
        if constexpr (PCS::kSupportsBatchMode && PCS::kQueryInstance) {
```

Reply:
I think this is more correct.

```suggestion
        if constexpr (PCS::kQueryInstance) {
          if constexpr (PCS::kSupportsBatchMode) {
            // BatchCommit
          } else {
            // Commit
          }
        }         
```

Reply:
@chokobole I've checked it. Please resolve this comment after you check as well!